File: api/notes/retrieve.js
import { createClient } from "@supabase/supabase-js";

const supabase = createClient(
    process.env.SUPABASE_URL,
    process.env.SUPABASE_ANON_KEY
);

export default async function handler(req, res) {
    if (req.method !== "POST") {
        return res.status(405).json({ error: "Method not allowed" });
    }

    const { otp } = req.body;
    if (!otp) return res.status(400).json({ error: "OTP required" });

    const { data, error } = await supabase
        .from("notes")
        .select("*")
        .eq("otp", otp)
        .single();

    if (error || !data) {
        return res.status(404).json({ error: "Invalid OTP" });
    }

    if (new Date() > new Date(data.expires_at)) {
        await supabase.from("notes").delete().eq("id", data.id);
        return res.status(410).json({ error: "OTP expired" });
    }

    await supabase.from("notes").delete().eq("id", data.id);
    res.status(200).json({ text: data.encrypted_text });
}

File: api/notes/retrieve.py
import json
from datetime import datetime, timezone
from cryptography.fernet import Fernet
from supabase import create_client
import os

def handler(request):
    try:
        SUPABASE_URL = os.environ["SUPABASE_URL"]
        SUPABASE_KEY = os.environ["SUPABASE_ANON_KEY"]
        FERNET_KEY = os.environ["FERNET_KEY"].encode()

        supabase = create_client(SUPABASE_URL, SUPABASE_KEY)
        fernet = Fernet(FERNET_KEY)

        body = json.loads(request.body)
        otp = body.get("otp")
        if not otp:
            return {
                "statusCode": 400,
                "headers": {"Content-Type": "application/json"},
                "body": json.dumps({"error": "OTP required"})
            }

        res = supabase.table("notes") \
            .select("id, encrypted_text, expires_at") \
            .eq("otp", otp) \
            .single() \
            .execute()

        if not res.data:
            return {
                "statusCode": 404,
                "headers": {"Content-Type": "application/json"},
                "body": json.dumps({"error": "Invalid OTP"})
            }

        expires_at = datetime.fromisoformat(res.data["expires_at"])
        if datetime.now(timezone.utc) > expires_at:
            supabase.table("notes").delete().eq("id", res.data["id"]).execute()
            return {
                "statusCode": 410,
                "headers": {"Content-Type": "application/json"},
                "body": json.dumps({"error": "OTP expired"})
            }

        text = fernet.decrypt(res.data["encrypted_text"].encode()).decode()
        supabase.table("notes").delete().eq("id", res.data["id"]).execute()

        return {
            "statusCode": 200,
            "headers": {"Content-Type": "application/json"},
            "body": json.dumps({"text": text})
        }

    except Exception as e:
        return {
            "statusCode": 500,
            "headers": {"Content-Type": "application/json"},
            "body": json.dumps({"error": str(e)})
        }
